build/builder.go

```diff
@@ -539,29 +583,100 @@ func (b *Builder) Finish() error {
 		return b.buildError
 	}
 
+	// map of temporary -> final names for all updated shards + shard metadata files
+	artifactPaths := make(map[string]string)
+	for tmp, final := range b.finishedShards {
+		artifactPaths[tmp] = final
+	}
+
+	oldShards := b.opts.FindAllShards()
+
+	if b.opts.IsDelta {
+		// Delta shard builds need to update FileTombstone and branch commit information for all
+		// existing shards
+		for _, shard := range oldShards {
+			repositories, indexMetadata, err := zoekt.ReadMetadataPathAlive(shard)
+			if err != nil {
+				b.buildError = err
+				continue
+			}
+
+			for _, r := range repositories {
+				if r.ID == b.opts.RepositoryDescription.ID {
+
+					if len(b.opts.ChangedOrRemovedFiles) > 0 && r.FileTombstones == nil {
+						r.FileTombstones = make(map[string]struct{})
+					}
+
+					for _, f := range b.opts.ChangedOrRemovedFiles {
+						r.FileTombstones[f] = struct{}{}
+					}
+
+					// TODO: Should we fail early here if we detect that the set of branch names
+					// in the builder options differs from what's inside the repository? Otherwise,
+					// the only error we'll get is a panic when we actually search the shard.
+					r.Branches = b.opts.RepositoryDescription.Branches
```

Comment:
Thoughts on this? I think the caller needs to check about what kind of index they'll need to do with IndexState() anyway, but adding another check here and setting `buildError` could add another layer of defense.

Reply:
Done, and I also added a test for this here: https://github.com/sourcegraph/zoekt/blob/4d5062705c456709eae412252591a18d281a85aa/build/builder_test.go#L427-L459. 

Let me know if you think the test itself is worth it

Reply:
The test is nice. I would keep it.